driver/src/main/com/mongodb/GridFSBucketImpl.java

```diff
@@ -0,0 +1,316 @@
+    public GridFSBucket withBucketName(final String bucketName) {
+        return new GridFSBucketImpl(database, bucketName, chunkSizeBytes, codecRegistry, readPreference, writeConcern, filesCollection,
+                chunksCollection, false);
+    }
+
+    @Override
+    public GridFSBucket withChunkSizeBytes(final int chunkSizeBytes) {
+        return new GridFSBucketImpl(database, bucketName, chunkSizeBytes, codecRegistry, readPreference, writeConcern, filesCollection,
+                chunksCollection, checkedIndexes);
+    }
+
+    @Override
+    public GridFSBucket withCodecRegistry(final CodecRegistry codecRegistry) {
+        return new GridFSBucketImpl(database, bucketName, chunkSizeBytes, codecRegistry, readPreference, writeConcern, filesCollection,
+                chunksCollection, checkedIndexes);
+    }
+
+    @Override
+    public GridFSBucket withReadPreference(final ReadPreference readPreference) {
+        return new GridFSBucketImpl(database, bucketName, chunkSizeBytes, codecRegistry, readPreference, writeConcern, filesCollection,
+                chunksCollection, checkedIndexes);
+    }
+
+    @Override
+    public GridFSBucket withWriteConcern(final WriteConcern writeConcern) {
+        return new GridFSBucketImpl(database, bucketName, chunkSizeBytes, codecRegistry, readPreference, writeConcern, filesCollection,
+                chunksCollection, checkedIndexes);
+    }
+
+    @Override
+    public GridFSUploadStream openUploadStream(final String filename) {
+        return openUploadStream(filename, new GridFSUploadOptions());
+    }
+
+    @Override
+    public GridFSUploadStream openUploadStream(final String filename, final GridFSUploadOptions options) {
+        int chunkSize = options.getChunkSizeBytes() == null ? chunkSizeBytes : options.getChunkSizeBytes();
+        CodecRegistry codecRegistryToUse = options.getCodecRegistry() == null ? codecRegistry : options.getCodecRegistry();
+        Bson metadata = options.getMetadata() == null ? new BsonDocument() : options.getMetadata();
+        BsonDocument metadataBsonDocument = metadata.toBsonDocument(BsonDocument.class, codecRegistryToUse);
+        checkCreateIndex();
+        return new GridFSUploadStreamImpl(filesCollection, chunksCollection, new ObjectId(), filename, chunkSize, metadataBsonDocument);
+    }
+
+    @Override
+    public ObjectId uploadFromStream(final String filename, final InputStream source) {
+        return uploadFromStream(filename, source, new GridFSUploadOptions());
+    }
+
+    @Override
+    public ObjectId uploadFromStream(final String filename, final InputStream source, final GridFSUploadOptions options) {
+        GridFSUploadStream uploadStream = openUploadStream(filename, options);
+        int chunkSize = options.getChunkSizeBytes() == null ? chunkSizeBytes : options.getChunkSizeBytes();
+        byte[] buffer = new byte[chunkSize];
+        int len;
+        try {
+            while ((len = source.read(buffer)) != -1) {
```

Comment:
Seems correct to me - as you continually read and write until end of stream.
Also checked guava's method that looks similar: https://github.com/google/guava/blob/master/guava/src/com/google/common/io/ByteStreams.java#L92